azure-pipelines.yml

```diff
@@ -0,0 +1,164 @@
+
+              [string] $versionRegex = '(?<=version: ).*'
+              [string] $pubspecContent = Get-Content -Path "$(Build.SourcesDirectory)/device_calendar/pubspec.yaml" -Raw
+
+              Write-Host " (i) Current version: $version"
+              Write-Host " (i) Original pubspec.yaml content: $pubspecContent"
+
+              [bool] $hasMatches = $pubspecContent -match $versionRegex
+
+              Write-Host " (i) Found matches to replace version? $hasMatches"
+              
+              [string] $newPubspecContent = $pubspecContent -replace $versionRegex, "$version-dev.$(Build.BuildID)"
+
+              Write-Host " (i) Updated pubspec.yaml content: $newPubspecContent"
+
+              $newPubspecContent | Set-Content -Path "$(Build.SourcesDirectory)/device_calendar/pubspec.yaml"
+
+        - task: PowerShell@2
+          displayName: 'Pre-release versioning - CHANGELOG.md'
+          inputs:
+            targetType: 'inline'
+            script: |
+              [string] $changelogContent = Get-Content -Path "$(Build.SourcesDirectory)/device_calendar/CHANGELOG.md" -Raw
+
+              Write-Host " (i) Current version: $env.currentVersion"
+              Write-Host " (i) Original CHANGELOG.md content: $changelogContent"
+
+              [string] $newChangelogContent = $changelogContent -replace $env:currentVersion, "$env:currentVersion-dev.$(Build.BuildID)"
+
+              Write-Host " (i) Updated CHANGELOG.md content: $newChangelogContent"
+
+              $newChangelogContent | Set-Content -Path "$(Build.SourcesDirectory)/device_calendar/CHANGELOG.md"
+
+        - task: CmdLine@2
+          displayName: 'Dry run publish'
+          inputs:
+            workingDirectory: '$(Build.SourcesDirectory)/device_calendar'
+            script: '$(flutterExecPath) pub publish --dry-run'
+
+        - task: CmdLine@2
+          displayName: 'Publish'
+          inputs:
+            workingDirectory: '$(Build.SourcesDirectory)/device_calendar'
+            script: '$(flutterExecPath) pub publish --force'
+            
+  # ----------- CD Production -----------
+  - stage: ReleaseProd
+    dependsOn: Build
+    condition: and(succeeded(), eq(variables['Build.SourceBranch'], 'refs/master'))
+    jobs: 
+     - job: Release
+       pool:
+        vmImage: 'windows-latest'
+       steps:
+        - task: FlutterInstall@0
+          displayName: 'Flutter install'
+          inputs:
+            channel: 'stable'
+            version: 'latest'
+
+        - task: CmdLine@2
+          displayName: 'Dry run publish'
+          inputs:
+            workingDirectory: '$(Build.SourcesDirectory)/device_calendar'
+            script: '$(flutterExecPath) pub publish --dry-run'
+
+        - task: CmdLine@2
+          displayName: 'Publish'
+          inputs:
+            workingDirectory: '$(Build.SourcesDirectory)/device_calendar'
+            script: '$(flutterExecPath) pub publish --force'
+
+        - task: PowerShell@2
```

Comment:
We should do this before doing the publish.
We should also query github to see if the tag exists already - if it does, the build should fail.

Reply:
@nickrandolph I've moved to run the github release before publish, but need to find out how i can fail the build if a tag exists already (current task doesn't support this, it just creates for the existing tag)

Reply:
@nickrandolph, used the new task and tested an actual release with an existing tag and it returned an error